javalib/src/main/scala/java/net/URLDecoder.scala

```diff
@@ -0,0 +1,46 @@
+package java.net
+
+import java.io.UnsupportedEncodingException
+import java.nio.{CharBuffer, ByteBuffer}
+import java.nio.charset.Charset
+
+object URLDecoder {
+  def decode(s: String, enc: String): String = {
+
+    if (!Charset.isSupported(enc)) throw new UnsupportedEncodingException(enc)
+
+    val len = s.length
+    val charsetDecoder = Charset.forName(enc).newDecoder()
+    val charBuffer = CharBuffer.allocate(len)
+
+    var i = 0
+    while (i < len) {
+      s.charAt(i) match {
+        case '+' =>
+          charBuffer.append(' ')
+          i += 1
+
+        case '%' if i + 3 < len =>
+          val chunks = s.substring(i).grouped(3).takeWhile(_.charAt(0) == '%')
+          try {
+            val hexValues = chunks.map(_.substring(1))
+            val bytes = hexValues.map(Integer.parseInt(_, 16).toByte).toArray
+            val decoded = charsetDecoder.decode(ByteBuffer.wrap(bytes))
+            charBuffer.append(decoded)
```

Comment:
This is going to be prohibitively slow for such an algorithm. In the Java lib, we tend to reasonably favor performance.

A few tips:
* (Lazily) allocate one `ByteBuffer` large enough to hold all possible bytes once (hence of size `len / 3`) and reuse it.
* Fill in the `ByteBuffer` directly from the `parseInt`s instead of allocating arrays all the time.
* Don't use `substring(1)` + `parseInt`. Directly use `charAt(i)` and `Character.digit(char, 16)` twice.
* Fill in the buffer with consecutive bytes using another `while` loop.
* Use the overload of `CharsetDecoder.decode` that takes an output `CharBuffer` to directly decode into `charBuffer` rather than allocating temporary `CharBuffer`s in the process. You'll have to `flush()` and `reset()` after calling `decode()` in that case; but since by construction you can't overflow, it's not more difficult than that.